### ISSUE TITLE
`InvalidCertificateSignatureType` encountered in TLS 1.3 handshake

### ISSUE THREAD

**Issue by mdcarr941** [rustls/rustls#1161]
I encountered a strange `InvalidCertificateSignatureType` when attempting to establish a connection between an instance of `rustls` running as a server and another instance running as a client. I wrote a test which reproduces the issue in a fork of this repository [here](https://github.com/mdcarr941/rustls/blob/main/rustls/tests/repro.rs).

Tracing through with the debugger, I found that the signature algorithm negotiated by the client and server was RSASSA-PSS with SHA-512, even though the Subject Public Key Info in the server certificate specifies RSASSA-PSS with SHA-256. Working backwards from `emit_certificate_verify_tls13` in `src/server/tls13.rs`, it seems that the `schemes` argument is determined using the `signature_algorithms` extension in the ClientHello and the algorithms supported by the server's current cipher suite and protocol version. I can't find anywhere that the Subject Public Key Info in the server's certificate is taken into account.

I may be mistaken, but if this is the case then this implementation is not in compliance with the spec. From section 4.1.3 of RFC 8446: "In addition, the signature algorithm MUST be compatible with the key in the sender's end-entity certificate."

It seems like the minimal way to address this is to filter the `schemes` argument using the algorithm in Subject Public Key Info, if the server is configured to use certificate authentication. This could be done in `handle_client_hello` just before calling `emit_certificate_verify`. I'd be down to write the PR but I'm not familiar with webpki, so if you could give me some pointers on parsing the server cert I'd appreciate it.

**Comment by ctz** [comment]
rustls currently only advertises PSS with:
```
          /* RSASSA-PSS algorithms with public key OID rsaEncryption */
          rsa_pss_rsae_sha256(0x0804),
          rsa_pss_rsae_sha384(0x0805),
          rsa_pss_rsae_sha512(0x0806),
```
and _not_:
```
          /* RSASSA-PSS algorithms with public key OID RSASSA-PSS */
          rsa_pss_pss_sha256(0x0809),
          rsa_pss_pss_sha384(0x080a),
          rsa_pss_pss_sha512(0x080b),
```

The first is the MTI option, and so I doubt that the latter three have much support in other libraries.

**Comment by mdcarr941** [comment]
Are you saying that certificates which specify RSASSA-PSS in their SPKI are not supported by rustls? The issue I have with using the rsaEncryption OID in my certificates is that I can't express constraints on key usage or message digest using them. Of course these constraints are only advisory, but at least when I specify RSASSA-PSS in my certificates a standards compliant application which doesn't understand them will error out rather than misusing the key.

**Comment by wqweto** [comment]
> I may be mistaken, but if this is the case then this implementation is not in compliance with the spec. From section 4.1.3 of RFC 8446: "In addition, the signature algorithm MUST be compatible with the key in the sender's end-entity certificate."

Btw, certificates with `Subject Public Key Info: Public Key Algorithm: rsaEncryption` (i.e. 1.2.840.113549.1.1.1) are compatibple with `rsa_pss_rsae_sha256`, `rsa_pss_rsae_sha384` and `rsa_pss_rsae_sha512` signatures (i.e. all hash sizes) and there is no hash info included in `Subject Public Key Info` for any reason.

The value of `Signature Algorithm` (e.g sha256WithRSAEncryption) identifies the type of signature used when the certificated was issued and does not limit the signature types its public key can be used for.

**Comment by ctz** [comment]
> @ctz Are you saying that certificates which specify RSASSA-PSS in their SPKI are not supported by rustls?

Correct.

> The issue I have with using the rsaEncryption OID in my certificates is that I can't express constraints on key usage or message digest using them. Of course these constraints are only advisory, but at least when I specify RSASSA-PSS in my certificates a standards compliant application which doesn't understand them will error out rather than misusing the key.

This error _is_ rustls erroring out rather than doing something else. But note that this is a parameter in the certificate, and only really alters how the public key is used. If you're wanting to control how and with what algorithms the private key is used (arguably the most important thing to worry about) then rustls has an extension point to allow control of this: https://docs.rs/rustls/latest/rustls/sign/trait.SigningKey.html .

**Comment by mdcarr941** [comment]
> This error is rustls erroring out rather than doing something else.
Yes, and I appreciate that, but it would be nice if rustls could support such certificates, provided that users of other types of certificates are not adversely affected. I do understand that this only affects the usage of the public key (I'm using the `SigningKey` trait to allow a private key stored in a TPM to be used to sign TLS messages, so thank you for that extension point).

Please consider the problem I'm facing. When I receive a peer certificate in a TLS handshake that only includes the rsaEncryption OID, I cannot determine the message digest the peer will use in its signature algorithm, I have to guess. I need to know this information so I can correctly verify the signature presented by the peer. Currently I'm only using SHA-256 in my tests, so its an easy guess. But, if I allow my users to specify the message digest, I'll run into issues. By using the id-RSASSA-PSS object in my certificates I can specify the message digest in its parameters field, thus eliminating this ambiguity.

Going back to my suggested change, I believe that filtering the `sigschemes_ex` in `handle_client_hello` using the scheme specified in the certificate's SPKI ensures that the signature algorithm agreed upon is compatible with the certificate. When doing this filtering I'll have to ensure that an SPKI with the rsaEncryption matches all available RSA signature schemes, while one which specifies id-RSASSA-PSS only matches the PSS variant for the message digest given in its parameters.

I believe that such a change is compatible with the RFC, based on the quote I gave in my opening post.

I suspect that this change won't break any users because if the `signature_algorithm` negotiated in the handshake differed from the one in their certificate, then signature verification would have failed. So if they're currently able to successfully handshake, and the filtering I implement selects the correct signature schemes, then no breakage will occur.

@ctz Is this a change you would be willing to review?
